Point_set_processing_3/include/CGAL/constraint_based_smooth_point_set.h

```diff
@@ -0,0 +1,652 @@
+    });
+
+  // update the normals
+  typedef boost::zip_iterator
+    <boost::tuple<iterator,
+                  typename std::vector<typename Kernel::Vector_3>::iterator> > Zip_iterator_4;
+
+  CGAL::for_each<ConcurrencyTag>
+    (CGAL::make_range (boost::make_zip_iterator (boost::make_tuple (points.begin(), new_normals.begin())),
+                      boost::make_zip_iterator (boost::make_tuple (points.end(), new_normals.end()))),
+    [&](const typename Zip_iterator_4::reference& t)
+    {
+      put (normal_map, get<0>(t), get<1>(t));
+      return true;
+    });
+
+  // early return if point smoothing is turned off
+  if(!do_point_smoothing)
+  {
+    return;
+  }
+
+  // compute the covariance matrix for each point
+  std::vector<std::pair<Eigen::Vector3d, Eigen::Matrix3d>> covm_optimized_eigens(nb_points);
+
+  CGAL::for_each<ConcurrencyTag>
+    (CGAL::make_range (boost::make_zip_iterator (boost::make_tuple
+                                                (points.begin(), pwns_neighbors.begin(), covm_optimized_eigens.begin())),
+                      boost::make_zip_iterator (boost::make_tuple
+                                                (points.end(), pwns_neighbors.end(), covm_optimized_eigens.end()))),
+    [&](const typename Zip_iterator_2::reference& t)
+    {
+      get<2>(t) = internal::calculate_covm_eigenvalues<Kernel, PointRange>
+          (get<0>(t),
+           point_map, normal_map,
+           get<1>(t),
+           normal_threshold,
+           eigenvalue_threshold_covm);
+      return true;
+    });
+
+  // optimize eigenvalues
+  internal::optimize_matrix_eigenvalues<Kernel>(covm_optimized_eigens, eigenvalue_threshold_covm);
+
+  // classify each point based on the cov matrix
+  std::vector<internal::point_type_t> point_classifications;
+  point_classifications.reserve(nb_points);
+
+  // for (auto it = nvt_optimized_eigens.begin() ; it < nvt_optimized_eigens.end() ; ++it)
+  for (auto it = covm_optimized_eigens.begin() ; it < covm_optimized_eigens.end() ; ++it)
+  {
+    point_classifications.push_back(internal::feature_detection<Kernel>(*it));
+  }
+
+  // colour the points FOR DEBUG ONLY
+  for (size_t i=0; i<point_classifications.size(); ++i)
+  {
+    switch(point_classifications[i])
+    {
+      case internal::point_type_t::corner:
+        get<2>(points[i]) = CGAL::make_array(static_cast<unsigned char>(255), static_cast<unsigned char>(0), static_cast<unsigned char>(0));
+        break;
+      case internal::point_type_t::edge:
+        get<2>(points[i]) = CGAL::make_array(static_cast<unsigned char>(0), static_cast<unsigned char>(255), static_cast<unsigned char>(0));
+        break;
+      case internal::point_type_t::flat:
+        get<2>(points[i]) = CGAL::make_array(static_cast<unsigned char>(0), static_cast<unsigned char>(0), static_cast<unsigned char>(255));
+        break;
+    }
+  }
+
+  // calculate diameter of point set
+  typedef CGAL::Min_sphere_of_points_d_traits_3<Kernel,FT> Traits;
```

Comment:
This introduces a heavy dependency. In case we have to keep it, we have to update package_info